[WebGPU] unsized arrays require bounds checking (251376)

mwyrzykowski opened the pull request:
#### 119b59ab84b5009acc5db1635ddc6d8281b8ff9d
<pre>
[WebGPU] unsized arrays require bounds checking (251376)
<a href="https://bugs.webkit.org/show_bug.cgi?id=251376">https://bugs.webkit.org/show_bug.cgi?id=251376</a>
&lt;radar://104827098&gt;

Reviewed by NOBODY (OOPS!).

Add the buffer length size to the argument buffer so the
generated metal source can perform bounds checking.

* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::makeBufferLengthDescriptorForBinding):
(WebGPU::Device::createBindGroupLayout):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::ComputePipeline::getBindGroupLayout):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::getBindGroupLayout):
* Websites/webkit.org/demos/webgpu/scripts/hello-cube-auto-layout.js:
* Websites/webkit.org/demos/webgpu/scripts/hello-cube-indirect.js:
* Websites/webkit.org/demos/webgpu/scripts/hello-indexed-cube.js:
* Websites/webkit.org/demos/webgpu/scripts/indirect-command-buffer-textured-cube.js:
* Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:
* Websites/webkit.org/demos/webgpu/scripts/query-sets.js:
* Websites/webkit.org/demos/webgpu/scripts/textured-cube-shader-constants.js:
* Websites/webkit.org/demos/webgpu/scripts/textured-cube-vs-input-buffers.js:
* Websites/webkit.org/demos/webgpu/scripts/textured-cube.js:
* Websites/webkit.org/demos/webgpu/scripts/two-cubes.js:
</pre>
----------------------------------------------------------------------
#### 67efd6e5cef85e43b8fc5a2738e437cad635809c
<pre>
WebGPU: Compute passes
<a href="https://bugs.webkit.org/show_bug.cgi?id=250863">https://bugs.webkit.org/show_bug.cgi?id=250863</a>
&lt;radar://83213548&gt;

Reviewed by NOBODY (OOPS!).

Implement compute passes which are needed to start
passing a lot of the CTS tests which create a simple
compute shader and read back the result, to ensure
the adapter was created.

<a href="https://bugs.webkit.org/show_bug.cgi?id=251171">https://bugs.webkit.org/show_bug.cgi?id=251171</a> has
been filed for tracking getting the threads per thread
group size from the shader compiler.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::beginRenderPass):

* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::dispatch):
(WebGPU::ComputePassEncoder::dispatchIndirect):
(WebGPU::ComputePassEncoder::setBindGroup):
(WebGPU::ComputePassEncoder::setPipeline):
(WebGPU::ComputePassEncoder::endPass):

* Source/WebGPU/WebGPU/BindGroupLayout.h:
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::BindGroupLayout::createEntryFromStructMember):
Moved so both compute and render pipeline can access.

* Source/WebGPU/WebGPU/ComputePipeline.h:
(WebGPU::ComputePipeline::create):
(WebGPU::ComputePipeline::threadsPerThreadgroup const):

* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::buildKeyValueReplacements const):
Moved to common path.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::getBindGroupLayout):
(WebGPU::buildKeyValueReplacements): Deleted.
(WebGPU::createEntryFromStructMember): Deleted.
Moved to common path.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/119b59ab84b5009acc5db1635ddc6d8281b8ff9d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105382 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14451 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38255 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114637 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15639 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5387 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97729 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114527 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111140 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12090 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95064 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39602 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93956 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26698 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81290 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7769 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28058 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7886 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4653 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13912 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47602 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9676 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WebGPU/BindGroup.mm:
```diff
@@ -108,6 +108,8 @@ static MTLRenderStages metalRenderStage(ShaderStage shaderStage)
 
                 [argumentEncoder[stage] setBuffer:buffer offset:entry.offset atIndex:index++];
+                ASSERT(limits().maxBufferSize < UINT32_MAX);
+                *(uint32_t*)[argumentEncoder[stage] constantDataAtIndex:index++] = static_cast<uint32_t>(entry.size);
```
Maybe we should do the signed one too while we're here? Shouldn't we be checking for overflow? We have `Checked` types for making that easy.

I think we should also be careful about units - any computation the shader has to perform will be executed millions of times per draw call, whereas any computation we can hoist to here will be executed once per bind group. If the value here is in "bytes" units, the shader will have to convert that to "indices" to bounds check against; if we can do that unit conversion here, that would probably be better in the end. (of course, that requires getting more info from the compiler about the size of the unsized array's inner type...)

Also, it sure would be good if there were a single shared function that would lower the BGL into a description of where everything in the Metal argument buffer will end up, rather than doing it like this via convention.

litherum replied:
oh, maybe we can't use the correct unit here because this code has to be able to run before any shaders are ever compiled

okay :|

mwyrzykowski replied:
based on some offline discussion I will implement the buffer lengths in a separate buffer

mwyrzykowski committed:
[WebGPU] unsized arrays require bounds checking (251376)
https://bugs.webkit.org/show_bug.cgi?id=251376
<radar://104827098>

Reviewed by NOBODY (OOPS!).

Add the buffer length size to the argument buffer so the
generated metal source can perform bounds checking.

* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createBindGroup):
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::makeBufferLengthDescriptorForBinding):
(WebGPU::Device::createBindGroupLayout):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::ComputePipeline::getBindGroupLayout):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::getBindGroupLayout):
* Websites/webkit.org/demos/webgpu/scripts/hello-cube-auto-layout.js:
* Websites/webkit.org/demos/webgpu/scripts/hello-cube-indirect.js:
* Websites/webkit.org/demos/webgpu/scripts/hello-indexed-cube.js:
* Websites/webkit.org/demos/webgpu/scripts/indirect-command-buffer-textured-cube.js:
* Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:
* Websites/webkit.org/demos/webgpu/scripts/query-sets.js:
* Websites/webkit.org/demos/webgpu/scripts/textured-cube-shader-constants.js:
* Websites/webkit.org/demos/webgpu/scripts/textured-cube-vs-input-buffers.js:
* Websites/webkit.org/demos/webgpu/scripts/textured-cube.js:
* Websites/webkit.org/demos/webgpu/scripts/two-cubes.js:

mwyrzykowski committed:
WebGPU: Compute passes
https://bugs.webkit.org/show_bug.cgi?id=250863
<radar://83213548>

Reviewed by NOBODY (OOPS!).

Implement compute passes which are needed to start
passing a lot of the CTS tests which create a simple
compute shader and read back the result, to ensure
the adapter was created.

https://bugs.webkit.org/show_bug.cgi?id=251171 has
been filed for tracking getting the threads per thread
group size from the shader compiler.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::beginRenderPass):

* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::dispatch):
(WebGPU::ComputePassEncoder::dispatchIndirect):
(WebGPU::ComputePassEncoder::setBindGroup):
(WebGPU::ComputePassEncoder::setPipeline):
(WebGPU::ComputePassEncoder::endPass):

* Source/WebGPU/WebGPU/BindGroupLayout.h:
* Source/WebGPU/WebGPU/BindGroupLayout.mm:
(WebGPU::BindGroupLayout::createEntryFromStructMember):
Moved so both compute and render pipeline can access.

* Source/WebGPU/WebGPU/ComputePipeline.h:
(WebGPU::ComputePipeline::create):
(WebGPU::ComputePipeline::threadsPerThreadgroup const):

* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::buildKeyValueReplacements const):
Moved to common path.

* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipeline):
(WebGPU::RenderPipeline::getBindGroupLayout):
(WebGPU::buildKeyValueReplacements): Deleted.
(WebGPU::createEntryFromStructMember): Deleted.
Moved to common path.
